<a href="https://colab.research.google.com/github/Shopping-Yuan/ML2021HW/blob/Shopping_vscode_branch/HW4/HW04_modified.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import shutil
shutil.copyfile('/content/drive/MyDrive/Dataset.zip','/content/Dataset.zip')

import zipfile
with zipfile.ZipFile('/content/Dataset.zip','r') as zip_ref:
  zip_ref.extractall('/content/')

In [ ]:
def parse_args():
  """arguments"""
  config = {
    "data_dir": "/content/Dataset",
    "save_path": "model.ckpt",
    "batch_size": 32,
    "n_workers": 8,
    "valid_steps": 2000,
    "warmup_steps": 1000,
    "save_steps": 10000,
    "total_steps": 70000,
  }

  return config

#Dataset

In [ ]:
import os
import json
import torch
import random
from pathlib import Path
from torch.utils.data import Dataset
from torch.nn.utils.rnn import pad_sequence


class myDataset(Dataset):
  def __init__(self, data_dir, segment_len=128):
    self.data_dir = data_dir
    self.segment_len = segment_len

    # Load the mapping from speaker neme to their corresponding id.
    # Path package can use / to join path
    #part 1-1
    mapping_path = Path(data_dir) / "mapping.json"
    mapping = json.load(mapping_path.open())
    self.speaker2id = mapping["speaker2id"]
    #self.speaker2id = {'id10001': 0, 'id10005': 1, ...... 'id11247': 598, 'id11251': 599}

    #part 1-2
    # Load metadata of training data.
    metadata_path = Path(data_dir) / "metadata.json"
    metadata = json.load(open(metadata_path))["speakers"]

    # Get the total number of speaker.
    self.speaker_num = len(metadata.keys())
    self.data = []
    for speaker in metadata.keys():
    #['id10473', 'id10328', 'id10382',......'id10864']
      for utterances in metadata[speaker]:
      # for example : metadata[speaker] = 'id10473'
      # utterances = {'feature_path': 'uttr-5c88b2f1803449789c36f14fb4d3c1eb.pt', 'mel_len': 652}......
        self.data.append([utterances["feature_path"], self.speaker2id[speaker]])
        # for example : ['uttr-5c88b2f1803449789c36f14fb4d3c1eb.pt',217]

  def __len__(self):
    return len(self.data)

  def __getitem__(self, index):
    feat_path, speaker = self.data[index]
    #[utterances["feature_path"], self.speaker2id[speaker]

    # Load preprocessed mel-spectrogram.
    # https://pytorch.org/docs/stable/generated/torch.load.html
    # using torch.load to open pt file . By default , tensors are loaded to GPU if they are GPU tensor.
    # part 2
    mel = torch.load(os.path.join(self.data_dir, feat_path))

    # Segmemt mel-spectrogram into "segment_len" frames.
    if len(mel) > self.segment_len:
      # Randomly get the starting point of the segment.
      start = random.randint(0, len(mel) - self.segment_len)
      # Get a segment with "segment_len" frames.
      mel = torch.FloatTensor(mel[start:start+self.segment_len])
    else:
      mel = torch.FloatTensor(mel)
    # Turn the speaker id into long for computing loss later.
    speaker = torch.FloatTensor([speaker]).long()
    return mel, speaker

  def get_speaker_number(self):
    return self.speaker_num

In [ ]:
#data_dir = parse_args()["data_dir"]
#part 1-1
#mapping.json checcking
#mapping_path = Path(data_dir) / "mapping.json"
#mapping = json.load(mapping_path.open())
#speaker2id = mapping["speaker2id"]
#print(type(mapping)) #<class 'dict'>
#print(mapping.keys()) #dict_keys(['speaker2id', 'id2speaker'])
#print(mapping.values()) #dict_values([{'id10001': 0, 'id10005': 1, ...... 'id11247': 598, 'id11251': 599},
             #{"0": "id10001", ......'599':  'id11251'}])
#print(mapping['speaker2id']['id10473']) #217

#part 1-2
#metadata.json checcking
#metadata_path = Path(data_dir) / "metadata.json"
#metadata = json.load(open(metadata_path))["speakers"]
#print(type(metadata)) #<class 'dict'>
#print(len(metadata.keys()),metadata.keys()) #600 dict_keys(['id10473', 'id10328', 'id10382',......'id10864'])
#print(len(metadata.values()),len(list(metadata.values())[0]),list(metadata.values())[0])
#600 176 [{'feature_path': 'uttr-5c88b2f1803449789c36f14fb4d3c1eb.pt', 'mel_len': 652},......
#{'feature_path': 'uttr-caf385f0032140e0add0bbde9a6595cd.pt', 'mel_len': 414}]
#print(len(metadata.values()),len(list(metadata.values())[1])) #600 101

#part 2
#mel = torch.load(os.path.join(data_dir, 'uttr-5c88b2f1803449789c36f14fb4d3c1eb.pt'))
#print(type(mel)) #<class 'torch.Tensor'>
#print(mel.size()) #torch.Size([652, 40])
#print(mel[0])
# tensor([
# 2.0060,  1.9015,  1.3473,  1.5711,  1.5569,  2.3200,  3.0064,  3.3848,
# 0.5376,  0.8397,  1.5710, -0.5662, -2.1262, -1.1271, -1.4272, -3.6233,
# -1.0765, -1.8453, -3.2415, -2.3427, -1.5888, -0.4692, -0.4291, -0.4722,
# -1.8450, -1.0638, -3.2177, -3.4006, -3.2176, -3.3409, -3.1142, -4.0395,
# -3.6502, -4.2407, -4.9656, -5.1238, -5.0620, -5.4308, -6.0757, -6.6186])

In [ ]:
import torch
from torch.utils.data import DataLoader, random_split
from torch.nn.utils.rnn import pad_sequence
#part 1 check collate_batch
def collate_batch(batch):
  # Process features within a batch.
  """Collate a batch of data."""
  #part 1-1 , 1-2
  mel, speaker = zip(*batch)
  # Because we train the model batch by batch, we need to pad the features in the same batch to make their lengths the same.
  mel = pad_sequence(mel, batch_first=True, padding_value=-20)    # pad log 10^(-20) which is very small value.
  # mel: (batch size, length, 40)
  #part 1-3
  return mel, torch.FloatTensor(speaker).long()


def get_dataloader(data_dir, batch_size, n_workers):
  """Generate dataloader"""
  dataset = myDataset(data_dir)
  speaker_num = dataset.get_speaker_number()
  # Split dataset into training dataset and validation dataset
  trainlen = int(0.9 * len(dataset))
  lengths = [trainlen, len(dataset) - trainlen]
  trainset, validset = random_split(dataset, lengths)

  train_loader = DataLoader(
    trainset,
    batch_size=batch_size,
    shuffle=True,
    drop_last=True,
    num_workers=n_workers,
    pin_memory=True,
    collate_fn=collate_batch,
  )
  valid_loader = DataLoader(
    validset,
    batch_size=batch_size,
    num_workers=n_workers,
    drop_last=True,
    pin_memory=True,
    collate_fn=collate_batch,
  )

  return train_loader, valid_loader, speaker_num

In [ ]:
# x = [["a",1],["b",2],["c",3]]
# print(list(zip(*x)))
# print(list(zip(("a",1),("b",2),("c",3))))

In [ ]:
#import numpy as np
#check collate_batch
#part 1-1
# * collect "sequence"function input to a tuple :
#def f(*paras):
#  print(paras)
#f(1,2,3) #(1, 2, 3)
# or collect a tuple to a "sequence"
#print(*(1,2,3)) #1 2 3

#part 1-2
#test_data = np.random.rand(2,3,4)
#test_label = np.random.rand(2,1)
#samples = (test_data,test_label)
#print(samples)
#x , y = zip((1,2),(3,4),(5,6),(7,8))
#print(x,y)

#part 1-3
#print(torch.FloatTensor([[1.],[1.]]))#tensor([[1.],[1.]])
#print(torch.FloatTensor([[1.],[1.]]).long())#tensor([[1],[1]]) implies the numbers are integer now
#print(torch.IntTensor([[1.],[1.]]))#tensor([[1],[1]], dtype=torch.int32)

In [ ]:
!pip install torchinfo

In [ ]:
# x = [True,True,False,True,False,False,False,True,False,True,True,False]

# true_index_list = []

# false_index_list = []

# for i,b in enumerate(x):
#   if b == True:
#     true_index_list += [i]
#   else :
#     false_index_list += [i]
# print(true_index_list,false_index_list)

# new_list = []
# t_max = len(true_index_list)
# f_max = len(false_index_list)
# true_index = 0
# false_index = 0
# t_or_f = 0
# while (t_max * f_max)>0:
#   if t_or_f % 2 ==0:
#     new_list += true_index_list[true_index:true_index+3]
#     true_index += 3
#     t_or_f += 1
#     t_max -= 3
#   else :
#     new_list += false_index_list[false_index:false_index+3]
#     false_index += 3
#     t_or_f += 1
#     f_max -= 3
# if t_max > 0:
#   new_list += true_index_list[true_index:]
# elif f_max > 0:
#   new_list += false_index_list[false_index:]

# print(new_list)


In [ ]:
import torch
import torch.nn as nn
import torch.nn.functional as F
import torchvision
from torchinfo import summary

class Classifier(nn.Module):
  def __init__(self, d_model=128, n_spks=600, dropout=0.1):
    super().__init__()
    # Project the dimension of features from that of input into d_model.
    self.prenet = nn.Linear(40, d_model)
    # TODO:
    #   Change Transformer to Conformer.
    #   https://arxiv.org/abs/2005.08100
    # embed_dim must be divisible by num_heads
    self.encoder_layer = nn.TransformerEncoderLayer(
      d_model=d_model, dim_feedforward=256, nhead=1
    )
    # batch_first = True
    # norm_first = True
    self.encoder = nn.TransformerEncoder(self.encoder_layer, num_layers=2)

    # Project the the dimension of features from d_model into speaker nums.
    self.pred_layer = nn.Sequential(
      nn.Linear(d_model, d_model),
      nn.ReLU(),
      nn.Linear(d_model, n_spks),
    )

  def forward(self, mels):
    """
    args:
      mels: (batch size, length, 40)
    return:
      out: (batch size, n_spks)
    """
    # out: (batch size, length, d_model)
    out = self.prenet(mels)
    # out: (length, batch size, d_model)
    out = out.permute(1, 0, 2)
    # The encoder layer expect features in the shape of (length, batch size, d_model).
    out = self.encoder_layer(out)
    # out: (batch size, length, d_model)
    out = out.transpose(0, 1)
    # mean pooling
    stats = out.mean(dim=1)

    # out: (batch, n_spks)
    out = self.pred_layer(stats)
    return out


In [ ]:
model = Classifier()
print(summary(model,(32,128,40)))
#print(model.state_dict().keys())
#print(model.state_dict()["encoder_layer.self_attn.in_proj_weight"].size())#torch.Size([240, 80]) (80+80+80)*80
#print(model.state_dict()["encoder_layer.self_attn.out_proj.weight"].size())#torch.Size([80, 80])

Layer (type:depth-idx)                   Output Shape              Param #
Classifier                               [32, 600]                 --
├─Linear: 1-1                            [32, 128, 80]             3,280
├─TransformerEncoderLayer: 1-2           [128, 32, 80]             --
│    └─MultiheadAttention: 2-1           [128, 32, 80]             25,920
│    └─Dropout: 2-2                      [128, 32, 80]             --
│    └─LayerNorm: 2-3                    [128, 32, 80]             160
│    └─Linear: 2-4                       [128, 32, 256]            20,736
│    └─Dropout: 2-5                      [128, 32, 256]            --
│    └─Linear: 2-6                       [128, 32, 80]             20,560
│    └─Dropout: 2-7                      [128, 32, 80]             --
│    └─LayerNorm: 2-8                    [128, 32, 80]             160
├─Sequential: 1-3                        [32, 600]                 --
│    └─Linear: 2-9                       [32, 80]                  6

parameter numner :
1.MultiheadAttention : =
d_model+1(bias) * d_query +
d_model+1(bias) * d_key  +
d_model+1(bias) * d_v   +
d_v + 1(bias) * output_dim( = dmodel)
= (80+1)80 + (80+1)80 + (80+1)80 + (80+1)80 = 25920

2.layer_norm : =input + input = 80*80 = 160




In [ ]:
import math

import torch
from torch.optim import Optimizer
from torch.optim.lr_scheduler import LambdaLR


def get_cosine_schedule_with_warmup(
  optimizer: Optimizer,
  num_warmup_steps: int,
  num_training_steps: int,
  num_cycles: float = 0.5,
  last_epoch: int = -1,
):
  """
  Create a schedule with a learning rate that decreases following the values of the cosine function between the
  initial lr set in the optimizer to 0, after a warmup period during which it increases linearly between 0 and the
  initial lr set in the optimizer.

  Args:
    optimizer (:class:`~torch.optim.Optimizer`):
      The optimizer for which to schedule the learning rate.
    num_warmup_steps (:obj:`int`):
      The number of steps for the warmup phase.
    num_training_steps (:obj:`int`):
      The total number of training steps.
    num_cycles (:obj:`float`, `optional`, defaults to 0.5):
      The number of waves in the cosine schedule (the defaults is to just decrease from the max value to 0
      following a half-cosine).
    last_epoch (:obj:`int`, `optional`, defaults to -1):
      The index of the last epoch when resuming training.

  Return:
    :obj:`torch.optim.lr_scheduler.LambdaLR` with the appropriate schedule.
  """

  def lr_lambda(current_step):
    # Warmup
    if current_step < num_warmup_steps:
      return float(current_step) / float(max(1, num_warmup_steps))
    # decadence
    # cosine angle
    progress = float(current_step - num_warmup_steps) / float(
      max(1, num_training_steps - num_warmup_steps)
    )
    return max(
      0.0, 0.5 * (1.0 + math.cos(math.pi * float(num_cycles) * 2.0 * progress))
    )

  return LambdaLR(optimizer, lr_lambda, last_epoch)

In [ ]:
from torch.optim import AdamW
optimizer = AdamW(model.parameters(), lr=1e-3)

scheduler = get_cosine_schedule_with_warmup(optimizer, 1000, 70000)

#print(type(scheduler.optimizer.param_groups)) #<class 'list'>
#print(len(scheduler.optimizer.param_groups))  #1
#print(type(scheduler.optimizer.param_groups[0])) #<class 'dict'>
#print(scheduler.optimizer.param_groups[0].keys())
#['params', 'lr', 'betas', 'eps', 'weight_decay', 'amsgrad', 'foreach',
#'maximize', 'capturable', 'differentiable', 'fused', 'initial_lr']
#print(scheduler.optimizer.param_groups[0]['params'])
#[Parameter containing: tensor([[......],requires_grad=True)]
#print(scheduler.optimizer.param_groups[0]['lr'])# 0.0
#print(scheduler.optimizer.param_groups[0]['betas'])# (0.9, 0.999)
#print(scheduler.optimizer.param_groups[0]['eps']) # 1e-08
#print(scheduler.optimizer.param_groups[0]['weight_decay']) #0.01
#print(scheduler.optimizer.param_groups[0]['amsgrad']) # False
#print(scheduler.optimizer.param_groups[0]['foreach']) # None

In [ ]:
import torch


def model_fn(batch, model, criterion, device):
  """Forward a batch through the model."""

  mels, labels = batch
  mels = mels.to(device)
  labels = labels.to(device)

  outs = model(mels)

  loss = criterion(outs, labels)

  # Get the speaker id with highest probability.
  preds = outs.argmax(1)
  # Compute accuracy.
  accuracy = torch.mean((preds == labels).float())

  return loss, accuracy

In [ ]:
from tqdm import tqdm
import torch


def valid(dataloader, model, criterion, device):
  """Validate on validation set."""

  model.eval()
  running_loss = 0.0
  running_accuracy = 0.0
  # Manually update the progress bar : step 1
  # total length
  pbar = tqdm(total=len(dataloader.dataset), ncols=0, desc="Valid", unit=" uttr")
  for i, batch in enumerate(dataloader):
    with torch.no_grad():
      loss, accuracy = model_fn(batch, model, criterion, device)
      running_loss += loss.item()
      running_accuracy += accuracy.item()
    # Manually update the progress bar : step 2
    # update each batch
    pbar.update(dataloader.batch_size)
    pbar.set_postfix(
      loss=f"{running_loss / (i+1):.2f}",
      accuracy=f"{running_accuracy / (i+1):.2f}",
    )
  # Manually update the progress bar : step 3
  pbar.close()
  model.train()

  return running_accuracy / len(dataloader)


In [ ]:
from tqdm import tqdm

import torch
import torch.nn as nn
from torch.optim import AdamW
from torch.utils.data import DataLoader, random_split


# def parse_args():
#   """arguments"""
#   config = {
#     "data_dir": "/content/Dataset",
#     "save_path": "model.ckpt",
#     "batch_size": 32,
#     "n_workers": 8,
#     "valid_steps": 2000,
#     "warmup_steps": 1000,
#     "save_steps": 10000,
#     "total_steps": 70000,
#   }

#   return config


def main(
  data_dir,
  save_path,
  batch_size,
  n_workers,
  valid_steps,
  warmup_steps,
  total_steps,
  save_steps,
):
  """Main function."""
  device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
  print(f"[Info]: Use {device} now!")

  train_loader, valid_loader, speaker_num = get_dataloader(data_dir, batch_size, n_workers)
  train_iterator = iter(train_loader)
  print(f"[Info]: Finish loading data!",flush = True)

  model = Classifier(n_spks=speaker_num).to(device)
  criterion = nn.CrossEntropyLoss()
  optimizer = AdamW(model.parameters(), lr=1e-3)
  scheduler = get_cosine_schedule_with_warmup(optimizer, warmup_steps, total_steps)
  print(f"[Info]: Finish creating model!",flush = True)

  best_accuracy = -1.0
  best_state_dict = None
  # pbar has total 2000 step
  pbar = tqdm(total=valid_steps, ncols=0, desc="Train", unit=" step")
  # print train loss and accuracy every 1000 step

  loss_acc_list = []

  for step in range(total_steps):
    # Get data
    try:
      batch = next(train_iterator)
    except StopIteration:
      train_iterator = iter(train_loader)
      batch = next(train_iterator)

    loss, accuracy = model_fn(batch, model, criterion, device)
    batch_loss = loss.item()
    batch_accuracy = accuracy.item()

    # Updata model
    loss.backward()
    optimizer.step()
    scheduler.step()
    optimizer.zero_grad()
    # print train loss acc
    loss_acc_list.append((loss, accuracy))
    print_loss , print_acc = list(zip(*loss_acc_list))
    if step % 1000 == 0:
      print(f"average loss = {sum(print_loss)/len(print_loss):.4f} \
      average accuracy = {sum(print_acc)/len(print_acc):.4f})")
      loss_acc_list = []

    # Log
    # update each step (as total)
    pbar.update()
    pbar.set_postfix(
      loss=f"{batch_loss:.2f}",
      accuracy=f"{batch_accuracy:.2f}",
      step=step + 1,
    )

    # Do validation
    # valid each 2000 step
    if (step + 1) % valid_steps == 0:
      #close above pbar
      pbar.close()

      valid_accuracy = valid(valid_loader, model, criterion, device)

      # keep the best model
      if valid_accuracy > best_accuracy:
        best_accuracy = valid_accuracy
        best_state_dict = model.state_dict()
      # new pbar has total 2000 step
      pbar = tqdm(total=valid_steps, ncols=0, desc="Train", unit=" step")

    # Save the best model so far.
    if (step + 1) % save_steps == 0 and best_state_dict is not None:
      torch.save(best_state_dict, save_path)
      pbar.write(f"Step {step + 1}, best model saved. (accuracy={best_accuracy:.4f})")

  pbar.close()


if __name__ == "__main__":
  main(**parse_args())

[Info]: Use cuda now!


/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py:557: UserWarning: This DataLoader will create 8 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  warnings.warn(_create_warning_msg(


[Info]: Finish loading data!
[Info]: Finish creating model!


Train:   0% 8/2000 [00:00<02:08, 15.55 step/s, accuracy=0.00, loss=6.37, step=8]

average loss = 6.4446       average accuracy = 0.0000)


Train:  50% 1002/2000 [00:53<00:50, 19.60 step/s, accuracy=0.03, loss=4.93, step=1002]

average loss = 5.6328       average accuracy = 0.0413)


Train: 100% 2000/2000 [01:54<00:00, 17.44 step/s, accuracy=0.28, loss=3.63, step=2000]
Valid: 100% 6944/6944 [00:14<00:00, 488.88 uttr/s, accuracy=0.23, loss=3.67]
Train:   0% 6/2000 [00:00<03:09, 10.55 step/s, accuracy=0.31, loss=3.57, step=2006]

average loss = 4.1010       average accuracy = 0.1747)


Train:  50% 1005/2000 [00:52<01:10, 14.09 step/s, accuracy=0.25, loss=3.38, step=3005]

average loss = 3.4500       average accuracy = 0.2649)


Train: 100% 2000/2000 [01:50<00:00, 18.03 step/s, accuracy=0.25, loss=3.63, step=4000]
Valid: 100% 6944/6944 [00:13<00:00, 506.90 uttr/s, accuracy=0.33, loss=3.13]
Train:   0% 4/2000 [00:00<03:41,  8.99 step/s, accuracy=0.34, loss=3.35, step=4004]

average loss = 3.1276       average accuracy = 0.3217)


Train:  50% 1004/2000 [00:53<01:06, 15.02 step/s, accuracy=0.19, loss=3.69, step=5004]

average loss = 2.9090       average accuracy = 0.3591)


Train: 100% 2000/2000 [01:52<00:00, 17.74 step/s, accuracy=0.53, loss=2.33, step=6000]
Valid: 100% 6944/6944 [00:13<00:00, 496.38 uttr/s, accuracy=0.40, loss=2.70]
Train:   0% 7/2000 [00:00<02:50, 11.68 step/s, accuracy=0.28, loss=2.94, step=6007]

average loss = 2.7517       average accuracy = 0.3871)


Train:  50% 1005/2000 [00:56<01:06, 15.06 step/s, accuracy=0.47, loss=2.16, step=7005]

average loss = 2.6014       average accuracy = 0.4109)


Train: 100% 2000/2000 [01:54<00:00, 17.47 step/s, accuracy=0.28, loss=3.08, step=8000]
Valid: 100% 6944/6944 [00:13<00:00, 516.37 uttr/s, accuracy=0.44, loss=2.48]
Train:   0% 9/2000 [00:00<02:29, 13.31 step/s, accuracy=0.47, loss=2.60, step=8009]

average loss = 2.4934       average accuracy = 0.4347)


Train:  50% 1006/2000 [00:57<01:03, 15.58 step/s, accuracy=0.56, loss=1.76, step=9006]

average loss = 2.3762       average accuracy = 0.4539)


Train: 100% 2000/2000 [01:52<00:00, 17.79 step/s, accuracy=0.50, loss=2.45, step=1e+4]
Valid: 100% 6944/6944 [00:13<00:00, 507.31 uttr/s, accuracy=0.46, loss=2.36]

Train:   0% 0/2000 [00:00<?, ? step/s]
Train:   0% 5/2000 [00:00<02:22, 14.03 step/s, accuracy=0.53, loss=1.94, step=1e+4]

Step 10000, best model saved. (accuracy=0.4579)
average loss = 2.2979       average accuracy = 0.4735)


Train:  50% 1002/2000 [00:57<00:53, 18.82 step/s, accuracy=0.53, loss=2.38, step=11002]

average loss = 2.1972       average accuracy = 0.4873)


Train: 100% 2000/2000 [01:51<00:00, 18.01 step/s, accuracy=0.62, loss=1.54, step=12000]
Valid: 100% 6944/6944 [00:13<00:00, 501.40 uttr/s, accuracy=0.48, loss=2.30]
Train:   1% 11/2000 [00:00<00:40, 49.59 step/s, accuracy=0.53, loss=2.38, step=12010]

average loss = 2.1264       average accuracy = 0.5051)


Train:  50% 1002/2000 [00:57<01:01, 16.21 step/s, accuracy=0.59, loss=1.88, step=13002]

average loss = 2.0546       average accuracy = 0.5163)


Train: 100% 2000/2000 [01:52<00:00, 17.76 step/s, accuracy=0.50, loss=2.04, step=14000]
Valid: 100% 6944/6944 [00:14<00:00, 495.33 uttr/s, accuracy=0.52, loss=2.14]
Train:   1% 11/2000 [00:00<00:39, 49.93 step/s, accuracy=0.62, loss=1.83, step=14010]

average loss = 2.0000       average accuracy = 0.5297)


Train:  50% 1002/2000 [00:57<00:51, 19.38 step/s, accuracy=0.72, loss=1.77, step=15002]

average loss = 1.9571       average accuracy = 0.5408)


Train: 100% 2000/2000 [01:48<00:00, 18.41 step/s, accuracy=0.62, loss=1.51, step=16000]
Valid: 100% 6944/6944 [00:14<00:00, 489.87 uttr/s, accuracy=0.53, loss=2.07]
Train:   0% 8/2000 [00:00<03:12, 10.36 step/s, accuracy=0.53, loss=2.46, step=16008]

average loss = 1.8760       average accuracy = 0.5537)


Train:  50% 1007/2000 [00:59<01:03, 15.69 step/s, accuracy=0.62, loss=2.35, step=17007]

average loss = 1.8337       average accuracy = 0.5639)


Train: 100% 2000/2000 [01:51<00:00, 17.98 step/s, accuracy=0.53, loss=1.86, step=18000]
Valid: 100% 6944/6944 [00:14<00:00, 481.64 uttr/s, accuracy=0.56, loss=1.93]
Train:   0% 6/2000 [00:00<03:15, 10.20 step/s, accuracy=0.72, loss=1.33, step=18006]

average loss = 1.8017       average accuracy = 0.5695)


Train:  50% 1007/2000 [01:01<01:00, 16.44 step/s, accuracy=0.56, loss=2.16, step=19007]

average loss = 1.7346       average accuracy = 0.5839)


Train: 100% 2000/2000 [01:56<00:00, 17.13 step/s, accuracy=0.41, loss=1.95, step=2e+4]
Valid: 100% 6944/6944 [00:14<00:00, 477.12 uttr/s, accuracy=0.57, loss=1.88]

Train:   0% 0/2000 [00:00<?, ? step/s]
Train:   0% 3/2000 [00:00<02:30, 13.23 step/s, accuracy=0.72, loss=1.15, step=2e+4]

Step 20000, best model saved. (accuracy=0.5667)
average loss = 1.7091       average accuracy = 0.5879)


Train:  50% 1002/2000 [01:01<00:52, 19.13 step/s, accuracy=0.72, loss=1.10, step=21002]

average loss = 1.6759       average accuracy = 0.5971)


Train: 100% 2000/2000 [01:50<00:00, 18.14 step/s, accuracy=0.56, loss=1.71, step=22000]
Valid: 100% 6944/6944 [00:14<00:00, 488.84 uttr/s, accuracy=0.58, loss=1.83]
Train:   1% 11/2000 [00:00<00:37, 52.56 step/s, accuracy=0.72, loss=1.30, step=22011]

average loss = 1.6171       average accuracy = 0.6116)


Train:  50% 1004/2000 [01:00<01:00, 16.56 step/s, accuracy=0.66, loss=1.26, step=23004]

average loss = 1.5914       average accuracy = 0.6175)


Train: 100% 2000/2000 [01:51<00:00, 17.96 step/s, accuracy=0.66, loss=1.34, step=24000]
Valid: 100% 6944/6944 [00:14<00:00, 471.77 uttr/s, accuracy=0.60, loss=1.76]
Train:   0% 8/2000 [00:00<02:50, 11.71 step/s, accuracy=0.66, loss=1.57, step=24008]

average loss = 1.5634       average accuracy = 0.6200)


Train:  50% 1006/2000 [01:01<00:58, 16.93 step/s, accuracy=0.69, loss=1.11, step=25006]

average loss = 1.5082       average accuracy = 0.6334)


Train: 100% 2000/2000 [01:50<00:00, 18.03 step/s, accuracy=0.78, loss=1.05, step=26000]
Valid: 100% 6944/6944 [00:14<00:00, 466.94 uttr/s, accuracy=0.60, loss=1.73]
Train:   0% 10/2000 [00:00<00:44, 44.88 step/s, accuracy=0.56, loss=2.07, step=26010]

average loss = 1.4751       average accuracy = 0.6381)


Train:  50% 1004/2000 [01:04<01:14, 13.29 step/s, accuracy=0.50, loss=2.06, step=27004]

average loss = 1.4803       average accuracy = 0.6392)


Train: 100% 2000/2000 [01:53<00:00, 17.61 step/s, accuracy=0.75, loss=1.05, step=28000]
Valid: 100% 6944/6944 [00:15<00:00, 460.65 uttr/s, accuracy=0.61, loss=1.71]
Train:   1% 11/2000 [00:00<00:38, 52.22 step/s, accuracy=0.69, loss=1.40, step=28011]

average loss = 1.4234       average accuracy = 0.6490)


Train:  50% 1003/2000 [01:06<01:06, 15.03 step/s, accuracy=0.69, loss=1.28, step=29003]

average loss = 1.4202       average accuracy = 0.6532)


Train: 100% 2000/2000 [01:56<00:00, 17.23 step/s, accuracy=0.72, loss=1.44, step=3e+4]
Valid: 100% 6944/6944 [00:15<00:00, 443.28 uttr/s, accuracy=0.62, loss=1.68]

Train:   0% 0/2000 [00:00<?, ? step/s]
Train:   0% 1/2000 [00:00<04:41,  7.11 step/s, accuracy=0.50, loss=1.79, step=3e+4]

Step 30000, best model saved. (accuracy=0.6151)
average loss = 1.3456       average accuracy = 0.6677)


Train:  50% 1003/2000 [01:05<01:02, 15.87 step/s, accuracy=0.62, loss=1.93, step=31003]

average loss = 1.3630       average accuracy = 0.6670)


Train: 100% 2000/2000 [01:53<00:00, 17.54 step/s, accuracy=0.66, loss=1.45, step=32000]
Valid: 100% 6944/6944 [00:15<00:00, 436.68 uttr/s, accuracy=0.63, loss=1.59]
Train:   0% 3/2000 [00:00<04:52,  6.84 step/s, accuracy=0.78, loss=1.04, step=32003]

average loss = 1.3040       average accuracy = 0.6778)


Train:  50% 1006/2000 [01:01<01:03, 15.61 step/s, accuracy=0.66, loss=1.18, step=33006]

average loss = 1.3093       average accuracy = 0.6759)


Train: 100% 2000/2000 [01:48<00:00, 18.39 step/s, accuracy=0.59, loss=1.33, step=34000]
Valid: 100% 6944/6944 [00:14<00:00, 478.15 uttr/s, accuracy=0.65, loss=1.53]
Train:   0% 6/2000 [00:00<02:50, 11.72 step/s, accuracy=0.66, loss=1.47, step=34006]

average loss = 1.2629       average accuracy = 0.6855)


Train:  50% 1006/2000 [01:02<01:07, 14.81 step/s, accuracy=0.69, loss=1.36, step=35006]

average loss = 1.2619       average accuracy = 0.6893)


Train: 100% 2000/2000 [01:50<00:00, 18.18 step/s, accuracy=0.69, loss=1.63, step=36000]
Valid: 100% 6944/6944 [00:14<00:00, 470.42 uttr/s, accuracy=0.65, loss=1.52]
Train:   0% 6/2000 [00:00<01:16, 25.98 step/s, accuracy=0.75, loss=0.96, step=36005]

average loss = 1.2119       average accuracy = 0.6972)


Train:  50% 1002/2000 [01:02<00:54, 18.24 step/s, accuracy=0.62, loss=1.93, step=37002]

average loss = 1.2144       average accuracy = 0.6982)


Train: 100% 2000/2000 [01:49<00:00, 18.31 step/s, accuracy=0.69, loss=1.04, step=38000]
Valid: 100% 6944/6944 [00:14<00:00, 464.26 uttr/s, accuracy=0.66, loss=1.50]
Train:   1% 11/2000 [00:00<00:40, 48.61 step/s, accuracy=0.72, loss=1.27, step=38011]

average loss = 1.1706       average accuracy = 0.7073)


Train:  50% 1002/2000 [01:03<00:57, 17.44 step/s, accuracy=0.69, loss=1.00, step=39002]

average loss = 1.1770       average accuracy = 0.7088)


Train: 100% 2000/2000 [01:51<00:00, 18.01 step/s, accuracy=0.72, loss=1.17, step=4e+4]
Valid: 100% 6944/6944 [00:15<00:00, 445.84 uttr/s, accuracy=0.66, loss=1.46]

Train:   0% 0/2000 [00:00<?, ? step/s]
Train:   0% 6/2000 [00:00<02:49, 11.76 step/s, accuracy=0.56, loss=2.13, step=4e+4]

Step 40000, best model saved. (accuracy=0.6613)
average loss = 1.1268       average accuracy = 0.7188)


Train:  50% 1003/2000 [01:06<01:52,  8.88 step/s, accuracy=0.78, loss=0.91, step=41003]

average loss = 1.1358       average accuracy = 0.7180)


Train: 100% 2000/2000 [01:52<00:00, 17.72 step/s, accuracy=0.81, loss=0.66, step=42000]
Valid: 100% 6944/6944 [00:15<00:00, 444.54 uttr/s, accuracy=0.68, loss=1.40]
Train:   0% 7/2000 [00:00<02:34, 12.89 step/s, accuracy=0.81, loss=0.82, step=42007]

average loss = 1.0904       average accuracy = 0.7279)


Train:  50% 1002/2000 [01:01<00:34, 29.24 step/s, accuracy=0.78, loss=0.71, step=43002]

average loss = 1.0836       average accuracy = 0.7277)


Train: 100% 2000/2000 [01:52<00:00, 17.85 step/s, accuracy=0.75, loss=0.96, step=44000]
Valid: 100% 6944/6944 [00:14<00:00, 464.08 uttr/s, accuracy=0.68, loss=1.38]
Train:   1% 11/2000 [00:00<00:39, 50.91 step/s, accuracy=0.81, loss=0.89, step=44011]

average loss = 1.0340       average accuracy = 0.7410)


Train:  50% 1001/2000 [01:01<00:45, 21.96 step/s, accuracy=0.84, loss=0.80, step=45001]

average loss = 1.0577       average accuracy = 0.7364)


Train: 100% 2000/2000 [01:50<00:00, 18.15 step/s, accuracy=0.88, loss=0.43, step=46000]
Valid: 100% 6944/6944 [00:15<00:00, 454.92 uttr/s, accuracy=0.69, loss=1.38]
Train:   1% 11/2000 [00:00<00:38, 51.04 step/s, accuracy=0.84, loss=1.07, step=46011]

average loss = 1.0145       average accuracy = 0.7468)


Train:  50% 1006/2000 [01:01<00:51, 19.33 step/s, accuracy=0.81, loss=0.83, step=47006]

average loss = 1.0052       average accuracy = 0.7494)


Train: 100% 2000/2000 [01:53<00:00, 17.65 step/s, accuracy=0.72, loss=0.83, step=48000]
Valid: 100% 6944/6944 [00:15<00:00, 447.45 uttr/s, accuracy=0.70, loss=1.32]
Train:   1% 11/2000 [00:00<00:39, 50.81 step/s, accuracy=0.72, loss=1.09, step=48011]

average loss = 0.9977       average accuracy = 0.7504)


Train:  50% 1005/2000 [00:58<00:57, 17.36 step/s, accuracy=0.81, loss=0.76, step=49005]

average loss = 0.9736       average accuracy = 0.7578)


Train: 100% 2000/2000 [01:48<00:00, 18.41 step/s, accuracy=0.75, loss=1.15, step=5e+4]
Valid: 100% 6944/6944 [00:15<00:00, 456.84 uttr/s, accuracy=0.71, loss=1.26]

Train:   0% 0/2000 [00:00<?, ? step/s]
Train:   0% 2/2000 [00:00<03:49,  8.70 step/s, accuracy=0.66, loss=1.50, step=5e+4]

Step 50000, best model saved. (accuracy=0.7058)
average loss = 0.9681       average accuracy = 0.7577)


Train:  50% 1001/2000 [00:56<00:45, 21.97 step/s, accuracy=0.78, loss=1.09, step=51001]

average loss = 0.9357       average accuracy = 0.7645)


Train: 100% 2000/2000 [01:48<00:00, 18.42 step/s, accuracy=0.81, loss=0.85, step=52000]
Valid: 100% 6944/6944 [00:15<00:00, 435.80 uttr/s, accuracy=0.71, loss=1.27]
Train:   0% 6/2000 [00:00<02:50, 11.67 step/s, accuracy=0.84, loss=0.68, step=52006]

average loss = 0.9240       average accuracy = 0.7697)


Train:  50% 1001/2000 [00:55<00:41, 24.25 step/s, accuracy=0.75, loss=1.22, step=53001]

average loss = 0.9263       average accuracy = 0.7688)


Train: 100% 2000/2000 [01:48<00:00, 18.46 step/s, accuracy=0.81, loss=0.68, step=54000]
Valid: 100% 6944/6944 [00:14<00:00, 466.80 uttr/s, accuracy=0.71, loss=1.26]
Train:   0% 10/2000 [00:00<02:11, 15.15 step/s, accuracy=0.72, loss=1.03, step=54010]

average loss = 0.9083       average accuracy = 0.7724)


Train:  50% 1001/2000 [00:54<00:36, 27.69 step/s, accuracy=0.78, loss=0.64, step=55001]

average loss = 0.8868       average accuracy = 0.7790)


Train: 100% 2000/2000 [01:47<00:00, 18.56 step/s, accuracy=0.78, loss=0.77, step=56000]
Valid: 100% 6944/6944 [00:14<00:00, 483.19 uttr/s, accuracy=0.72, loss=1.25]
Train:   0% 7/2000 [00:00<02:26, 13.61 step/s, accuracy=0.75, loss=1.00, step=56007]

average loss = 0.8765       average accuracy = 0.7797)


Train:  50% 1001/2000 [00:53<00:45, 21.90 step/s, accuracy=0.84, loss=0.55, step=57001]

average loss = 0.8848       average accuracy = 0.7778)


Train: 100% 2000/2000 [01:52<00:00, 17.80 step/s, accuracy=0.78, loss=0.62, step=58000]
Valid: 100% 6944/6944 [00:16<00:00, 420.51 uttr/s, accuracy=0.71, loss=1.24]
Train:   1% 13/2000 [00:00<00:34, 57.66 step/s, accuracy=0.84, loss=0.50, step=58013]

average loss = 0.8676       average accuracy = 0.7810)


Train:  50% 1002/2000 [00:54<00:38, 26.03 step/s, accuracy=0.78, loss=0.83, step=59002]

average loss = 0.8581       average accuracy = 0.7840)


Train: 100% 2000/2000 [01:52<00:00, 17.84 step/s, accuracy=0.78, loss=0.92, step=6e+4]
Valid: 100% 6944/6944 [00:14<00:00, 478.18 uttr/s, accuracy=0.72, loss=1.21]

Train:   0% 0/2000 [00:00<?, ? step/s]
Train:   0% 7/2000 [00:00<02:40, 12.45 step/s, accuracy=0.88, loss=0.43, step=6e+4]

Step 60000, best model saved. (accuracy=0.7167)
average loss = 0.8612       average accuracy = 0.7863)


Train:  50% 1002/2000 [00:53<00:51, 19.47 step/s, accuracy=0.78, loss=0.86, step=61002]

average loss = 0.8367       average accuracy = 0.7908)


Train: 100% 2000/2000 [01:52<00:00, 17.75 step/s, accuracy=0.88, loss=0.53, step=62000]
Valid: 100% 6944/6944 [00:14<00:00, 475.59 uttr/s, accuracy=0.73, loss=1.22]
Train:   1% 11/2000 [00:00<00:36, 53.96 step/s, accuracy=0.81, loss=0.70, step=62011]

average loss = 0.8364       average accuracy = 0.7875)


Train:  50% 1003/2000 [00:53<00:54, 18.17 step/s, accuracy=0.75, loss=0.73, step=63003]

average loss = 0.8339       average accuracy = 0.7922)


Train: 100% 2000/2000 [01:52<00:00, 17.72 step/s, accuracy=0.59, loss=1.62, step=64000]
Valid: 100% 6944/6944 [00:15<00:00, 460.85 uttr/s, accuracy=0.72, loss=1.20]
Train:   0% 10/2000 [00:00<00:40, 49.07 step/s, accuracy=0.81, loss=0.53, step=64010]

average loss = 0.8230       average accuracy = 0.7911)


Train:  50% 1001/2000 [00:54<00:44, 22.47 step/s, accuracy=0.84, loss=0.58, step=65001]

average loss = 0.8215       average accuracy = 0.7942)


Train: 100% 2000/2000 [01:55<00:00, 17.33 step/s, accuracy=0.81, loss=1.17, step=66000]
Valid: 100% 6944/6944 [00:14<00:00, 487.68 uttr/s, accuracy=0.73, loss=1.18]
Train:   0% 7/2000 [00:00<01:04, 30.87 step/s, accuracy=0.88, loss=0.38, step=66007]

average loss = 0.8053       average accuracy = 0.7963)


Train:  50% 1003/2000 [00:51<01:09, 14.43 step/s, accuracy=0.81, loss=0.88, step=67003]

average loss = 0.8081       average accuracy = 0.7983)


Train:  87% 1747/2000 [01:33<00:17, 14.64 step/s, accuracy=0.75, loss=0.71, step=67747]

In [ ]:
import os
import json
import torch
from pathlib import Path
from torch.utils.data import Dataset


class InferenceDataset(Dataset):
  def __init__(self, data_dir):
    # using / to merge path
    testdata_path = Path(data_dir) / "testdata.json"
    metadata = json.load(testdata_path.open())
    self.data_dir = data_dir
    #
    self.data = metadata["utterances"]

  def __len__(self):
    return len(self.data)

  def __getitem__(self, index):
    utterance = self.data[index]
    feat_path = utterance["feature_path"]
    mel = torch.load(os.path.join(self.data_dir, feat_path))
    # compare : when we load training set, we use append to concate datas
    return feat_path, mel


def inference_collate_batch(batch):
  """Collate a batch of data."""
  feat_paths, mels = zip(*batch)

  return feat_paths, torch.stack(mels)

In [ ]:
import json
import csv
from pathlib import Path
from tqdm.notebook import tqdm

import torch
from torch.utils.data import DataLoader

def parse_args():
  """arguments"""
  config = {
    "data_dir": "./Dataset",
    "model_path": "./model.ckpt",
    "output_path": "./output.csv",
  }

  return config


def main(
  data_dir,
  model_path,
  output_path,
):
  """Main function."""
  device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
  print(f"[Info]: Use {device} now!")

  mapping_path = Path(data_dir) / "mapping.json"
  mapping = json.load(mapping_path.open())

  dataset = InferenceDataset(data_dir)
  dataloader = DataLoader(
    dataset,
    batch_size=1,
    shuffle=False,
    drop_last=False,
    num_workers=8,
    collate_fn=inference_collate_batch,
  )
  print(f"[Info]: Finish loading data!",flush = True)

  speaker_num = len(mapping["id2speaker"])
  model = Classifier(n_spks=speaker_num).to(device)
  model.load_state_dict(torch.load(model_path))
  model.eval()
  print(f"[Info]: Finish creating model!",flush = True)

  results = [["Id", "Category"]]
  for feat_paths, mels in tqdm(dataloader):
    with torch.no_grad():
      mels = mels.to(device)
      outs = model(mels)
      preds = outs.argmax(1).cpu().numpy()
      # zip from two list to a sequence of tuples
      # preds is a numpy array whose elements are int numbers
      for feat_path, pred in zip(feat_paths, preds):
        # mapping["id2speaker"] = {"0": "id10001", ......'599':  'id11251'}])
        results.append([feat_path, mapping["id2speaker"][str(pred)]])

  with open(output_path, 'w', newline='') as csvfile:
    writer = csv.writer(csvfile)
    writer.writerows(results)


if __name__ == "__main__":
  main(**parse_args())


[Info]: Use cuda now!
[Info]: Finish loading data!
[Info]: Finish creating model!


  0%|          | 0/6000 [00:00<?, ?it/s]